__tests__/CognitoUtils.test.js

```diff
@@ -0,0 +1,148 @@
+      return CognitoUtils.authenticateUser(user, password)
+        .then(authResult => {
+          expect(authUserSpy).toHaveBeenCalledTimes(1)
+          expect(authResult).toEqual(mockSession)
+        })
+        .catch(error => { throw error })
+    })
+
+    it('rejects on failed authentication', () => {
+      const mockError = {}
+      const authUserSpy = jest.spyOn(user, 'authenticateUser').mockImplementation((_details, callback) => {
+        return callback.onFailure(mockError)
+      })
+      return CognitoUtils.authenticateUser(user, password)
+        .then(result => { throw result })
+        .catch(error => {
+          expect(authUserSpy).toHaveBeenCalledTimes(1)
+          expect(error).toEqual(mockError)
+        })
+    })
+  })
+
+  describe('getSession()', () => {
+    const user = CognitoUtils.cognitoUser(username)
+
+    afterEach(() => {
+      jest.restoreAllMocks()
+    })
+
+    it('returns a promise', () => {
+      const result = CognitoUtils.getSession(user)
+      expect(result).toBeInstanceOf(Promise)
+    })
+
+    it('resolves with session data when successful', () => {
+      const mockError = null
+      const mockSession = {}
+      const getSessionSpy = jest.spyOn(user, 'getSession').mockImplementation(callback => {
+        return callback(mockError, mockSession)
+      })
+      return CognitoUtils.getSession(user)
+        .then(result => {
+          expect(getSessionSpy).toHaveBeenCalledTimes(1)
+          expect(result).toEqual(mockSession)
+        })
+        .catch(error => { throw error })
+    })
+
+    it('rejects with error info when failing', () => {
+      const mockError = 'uh oh!'
+      const getSessionSpy = jest.spyOn(user, 'getSession').mockImplementation(callback => {
+        return callback(mockError)
+      })
+      return CognitoUtils.getSession(user)
+        .then(result => { throw result })
+        .catch(error => {
+          expect(getSessionSpy).toHaveBeenCalledTimes(1)
+          expect(error).toEqual(mockError)
+        })
+    })
+  })
+
+  describe('getIdTokenString()', () => {
+    const user = CognitoUtils.cognitoUser(username)
+
+    afterEach(() => {
+      jest.restoreAllMocks()
+    })
+
+    it('returns a promise', () => {
+      const result = CognitoUtils.getIdTokenString(user)
+      expect(result).toBeInstanceOf(Promise)
+    })
+
+    it('resolves to a JSON web token string', () => {
+      const token = 'eyjqew5u4yijrg09jh40ghi909tj9jgh...'
+      const mockError = null
+      const mockSession = {
+        idToken: {
+          jwtToken: token
+        }
+      }
+      const getSessionSpy = jest.spyOn(user, 'getSession').mockImplementation(callback => {
+        return callback(mockError, mockSession)
+      })
+      return CognitoUtils.getIdTokenString(user)
+        .then(result => {
+          expect(getSessionSpy).toHaveBeenCalledTimes(1)
+          expect(result).toEqual(token)
+        })
+        .catch(error => { throw error })
+    })
```

Comment:
test for the failure case might be nice, but wouldn't hold up merging over it.

Reply:
Would be good to add. What would/should that function do when there's a failure? I don't believe it catches.

Reply:
i think just bubble up the original error, though i hadn't thought about it much. we'll see if a test that effect ends up exposing weird behavior or something.